backend-docker/server/api.go

```diff
@@ -163,17 +168,35 @@ func (s *Server) buildImageHandler(w http.ResponseWriter, r *http.Request) {
 			Response{w}.ServerError("while dumping file part ", err)
 			return
 		}
-	}
 
-	if _, err := os.Stat(filepath.Join(buildDir, "Dockerfile")); os.IsNotExist(err) {
-		Response{w}.ServerError("no Dockerfile to build new image ", err)
-		return
+		if strings.HasSuffix(strings.ToLower(part.FileName()), ".tar") || strings.HasSuffix(strings.ToLower(part.FileName()), ".tar.gz") {
```

Comment:
I think the logic is not perfect here. I would first save all the files on the file system, and then 1. if there exists a Dockerfile, assume build image is expected, or 2. if there is a single tar file, assume an image.

Right now the user may send a Dockerfile together with a .tar file that is used in the Dockerfile, and the assumption that the .tar file is an image would be incorrect.